### PR TITLE
Add pydantic core uuid

### PR DESCRIPTION
**Pull request by JeanArhancet**
## Change Summary

Integration test for this MR : https://github.com/pydantic/pydantic-core/pull/772

## Checklist

* [ ] The pull request title is a good summary of the changes - it will be used in the changelog
* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

**Comment by davidhewitt** [comment]
Naturally it fails to build 😂. I think it would be fixed by installing a rust toolchain, similar to this: https://github.com/pydantic/pydantic-core/pull/782

**Comment by JeanArhancet** [comment]
@davidhewitt  Do you have an idea of the CI issue?

**Comment by davidhewitt** [comment]
Not off the top of my head, I'll follow up on this (probably tomorrow at this point, I got pulled sideways onto #6727).

**Comment by JeanArhancet** [comment]
I had to make a lot of changes but I don't think it's any good: 

- `UUID` -> `uuid` in error messages
- `class` -> `class_name` in error messages
-  Add a field `version` in metadata
- ...

Also, I've changed the UUID1...etc types

@davidhewitt WDYT?

File: pydantic/json_schema.py [hunk 1]
```diff
@@ -1630,6 +1630,17 @@ def url_schema(self, schema: core_schema.UrlSchema) -> JsonSchemaValue:
...
+    def uuid_schema(self, schema: core_schema.UuidSchema) -> JsonSchemaValue:
+        """Generates a JSON schema that matches a schema that defines a UUID.
+
+        Args:
+            schema: The core schema.
+
+        Returns:
+            The generated JSON schema.
+        """
+        return {'type': 'string', 'format': 'uuid'}
```
Comment (davidhewitt) [inline pydantic/json_schema.py:1642]: I think we might want to have the uuid version in the format (where applicable)?

File: tests/test_json.py [hunk 2]
```diff
@@ -48,7 +48,7 @@ class MyModel(BaseModel):
 @pytest.mark.parametrize(
     'ser_type,gen_value,json_output',
     [
-        (UUID, lambda: 'ebcdab58-6eb8-46fb-a190-d07a33e9eac8', b'"ebcdab58-6eb8-46fb-a190-d07a33e9eac8"'),
+        (UUID, lambda: UUID('ebcdab58-6eb8-46fb-a190-d07a33e9eac8'), b'"ebcdab58-6eb8-46fb-a190-d07a33e9eac8"'),
```
Comment (davidhewitt) [inline tests/test_json.py:51]: Why was this change necessary?

File: pydantic/types.py [hunk 3]
```diff
@@ -527,46 +528,36 @@ def condecimal(
...
+
+    version: int
 
     def __hash__(self) -> int:
-        return hash(type(self.uuid_version))
+        return hash(self.version)
 
 
+UuidVersion = Annotated[int, Version]
+UUIDStrict = Annotated[UUID, Strict()]
```
Comment (davidhewitt) [inline pydantic/types.py:555]: Personally I don't think we need to add this, as users already have `Annotated[UUID, Strict()]` recommended as https://docs.pydantic.dev/latest/usage/strict_mode/#strict-mode-with-annotated-strict

If we added this, users would also want `UUID1Strict` etc presumably, and it all gets a bit too much then IMO.

File: pydantic/types.py [hunk 4]
```diff
@@ -527,46 +528,36 @@ def condecimal(
...
+
+        UuidVersion = Annotated[int, Version]
+        ```
+    """
+
+    version: int
 
     def __hash__(self) -> int:
-        return hash(type(self.uuid_version))
+        return hash(self.version)
```
Comment (davidhewitt) [inline pydantic/types.py:551]: cc @adriangb I would be grateful to hear your input here on what you is appropriate to do with this type. My instinct is against adding a generic `Version` type and keeping it just as `UuidVersion`, maybe there's a technical reason I've missed for doing this. We should be careful not to create a breaking change.

**Review by davidhewitt — COMMENTED** [review]
Thanks, this is helpful to see. I've spattered a few questions on the diff here.

I also think we probably do not want to have the `uuid_exact_type` error, to keep things simpler. I've posted a corresponding suggestion on the `pydantic-core` PR.

I got diverted by memory leaks yesterday, I am now returning to the decimal validator so will figure out the CI issue here and the benchmark example ASAP.

File: tests/test_json.py (hunk 2 above)
Comment (JeanArhancet) [reply tests/test_json.py:51]: To match the new UUID serializer https://github.com/pydantic/pydantic-core/pull/772/files#diff-3e8029d3656b4c545330496a646d8b651884a649d1a7934fd7f32c50e33fe8a7R10-R20

Do you think it's a regression?

File: tests/test_types.py [hunk 5]
```diff
@@ -1288,7 +1289,7 @@ def __bool__(self) -> bool:
         ('float_check', b't', ValidationError),
         ('uuid_check', 'ebcdab58-6eb8-46fb-a190-d07a33e9eac8', UUID('ebcdab58-6eb8-46fb-a190-d07a33e9eac8')),
         ('uuid_check', UUID('ebcdab58-6eb8-46fb-a190-d07a33e9eac8'), UUID('ebcdab58-6eb8-46fb-a190-d07a33e9eac8')),
-        ('uuid_check', b'ebcdab58-6eb8-46fb-a190-d07a33e9eac8', UUID('ebcdab58-6eb8-46fb-a190-d07a33e9eac8')),
+        ('uuid_check', b'ebcdab58-6eb8-46fb-a190-d07a33e9eac8', ValidationError),
```
Comment (JeanArhancet) [inline tests/test_types.py:1292]: If I run a python script to test, I get an error. So I think it's normal to have a `ValidatorError`

```print(str(UUID(b'a6cc5730-2261-11ee-9c43-2eb5a363657c')))``` -> `TypeError: a bytes-like object is required, not 'str'`

**Comment by JeanArhancet** [comment]
@davidhewitt I went back on the changes

File: pydantic/types.py (hunk 3 above)
Comment (JeanArhancet) [reply pydantic/types.py]: You're right, thanks

File: pydantic/json_schema.py (hunk 1 above)
Comment (JeanArhancet) [reply pydantic/json_schema.py:1642]: I think we deduplicate the information because if we have a uuid4 the json is : ` {'type': 'string', 'format': 'uuid4'}`

**Comment by JeanArhancet** [comment]
Sorry to ask again @davidhewitt 

But I seem to have a regression here: 

```
│  tests/test_types.py  │  test_union_subclass[10]    │  4706           │  4715        │  AssertionError  │
│  tests/test_types.py  │  test_union_subclass[None]  │  4706           │  4715        │  AssertionError  │
```

I don't understand why. Do you have any idea ?

**Comment by JeanArhancet** [comment]
superseded by #6831
